### DISCUSSION TITLE
Design for Position UI

### DISCUSSION THREAD

**Discussion by jameskoster** [WordPress/gutenberg#46032] [Show and tell]
CSS position has many permutations, and while we're unlikely to implement everything in one go it would be good to settle on a design that can serve us equally well in the short and long term. 

`Sticky` (and potentially `fixed`) are currently being worked on in https://github.com/WordPress/gutenberg/pull/46142 (originally started in: https://github.com/WordPress/gutenberg/pull/44723). 

Here's a riff on a design shared earlier by @jasmussen:

![Fixed](https://user-images.githubusercontent.com/846565/203781500-004bf0ba-890b-482a-833d-b76524def366.png)

* Position control lives in the 'Advanced' panel. Previous explorations had it in the 'Layout' panel but I'm not sure it's a good fit:
  1. Only containers have a Layout panel, and it may be that you want to fix the position of a single block.
  2. All the controls in the Layout panel currently relate explicitly to child blocks, and position does not.
  3. Layout controls change based on whether the container has flow or flex layout, while position has no relation to that condition.
* A dropdown contains the different layout options, and based on the selection secondary controls are exposed.
* `Fixed` selection includes a matrix picker for easy position selection, and a Settings icon which exposes more granular control.
* `Sticky` selection includes a Settings icon where you can specify offset positioning along the X / Y axis.
* In List View, fixed-position blocks are separated, which essentially communicates the z-index value.

In the future we can add:

* `Absolute` selection, which would have the same secondary tools as `Fixed`.
* `Relative` selection which would not require any secondary tools:

<img width="738" alt="Screenshot 2022-11-24 at 12 21 34" src="https://user-images.githubusercontent.com/846565/203783401-f9959785-b1c7-4dba-aec1-612663dac71c.png">

Let's discuss.

**Comment by ramonjd** [comment]
> In List View, fixed-position blocks are separated, which essentially communicates the z-index value.

Thanks for working on the designs! 

How do you think it could work with, say, nested Groups that have fixed position enabled? I'm coming from the assumption that the list view represents a tree of elements. 

The design does make elements sitting "on top" more accessible, but on the face of it, splitting the tree into fixed and scroll appears to me to diverge from the tree paradigm.

Just thinking out loud here: it would be ace to be able to switch the list view between a strict hierarchy (what it is now) and some representation of z-index order. Or not 😄 🤷  

Or, in the opposite direction, are visual aids required at all in the list view given that a user can click on the sticky element and edit the properties?

**Reply by jameskoster** [reply]
> are visual aids required at all in the list view given that a user can click on the sticky element and edit the properties?

I think it helps to quickly explain why the position of a block on the canvas is not reflecting the tree structure. But I agree we'll need to explore other options, as the separation doesn't work for nested fixed blocks.

**Comment by andrewserong** [comment]
> Position control lives in the 'Advanced' panel.

Nice reasoning here! I think that makes for a really good pragmatic step for introducing the control, and good work honing in on all the conceptual details surrounding layout. I was chatting with @tellthemachines about this too, and I think you've made some great arguments for decoupling the concept of Position from Layout. In terms of `block.json` we can represent it as `supports.position` as its own block support potentially, rather than tying it to layout under `supports.__experimentalLayout.allowPosition`, which I've been doing in my explorations up until now.

Also, the idea of allowing relative + absolute positioning further down the track will unlock some very creative possibilities for laying things out with blocks, so that's another good argument for decoupling from container blocks (e.g. set a Group block to "relative" and then set a child Image block to "absolute" and nudge it off to the side a little for a deliberate aesthetic) 👍

> In List View, fixed-position blocks are separated, which essentially communicates the z-index value.

I imagine moving things around in the list view based on position attributes could be tricky to implement. Though I think it's a good thing to explore, would this be a blocker for the initial implementation? I.e. in an initial implementation with sticky and top position, would it be acceptable if it weren't flagged in the list view, or do you think having some visual indication there is a blocker for usability?

Thanks again for splitting out this discussion! 🙂

**Reply by jameskoster** [reply]
Thanks for the words of encouragement!

> would this be a blocker for the initial implementation?

No, certainly not a blocker :)

**Comment by tellthemachines** [comment]
Thanks for starting this discussion! I have a few thoughts to add:

### Controls and implementation

With "fixed", "sticky", "absolute" and "relative" it seems increasingly like we are trying to implement a UI for CSS functionality. Is this really what we want? It might be good to take a step back, consider the needs of our users (not all of whom are familiar with CSS) and build from there, rather than taking the implementation as a starting point and building the tools around the CSS properties we can expose.

CSS `position: sticky` in particular is a tricky one. If we add a control that just toggles that property on a block, chances are it won't work as users expect it to, because in order to stick, the block needs to be in a container taller than itself and and it will only stick as far as its container is still on screen. This means that if a template is sectioned into areas, the sticky container will need to be in the outermost wrapper in order to stay sticky across the whole page.

 We can solve this either with detailed instructions on how to make it work, or by making it work magically with JS. I suspect from a user point of view, a "magically working" experience will be much better.

### List View

With the separation of positioned blocks into their own section in List View, we need to consider accessibility very carefully. Currently, List View is the most accessible way to navigate the editor. Its intent to mirror the document structure is crucial particularly when the interface is visually complex enough that we can no longer tell at a glance how many blocks we have on the page.

I anticipate that adding fixed or otherwise positioned blocks will only make the canvas harder to navigate and edit, and if List View no longer reflects the DOM order that can make tasks like moving blocks much harder. We should particularly consider that positioning often depends on a parent element (whether a sticky container, or a relatively positioned ancestor to an absolutely positioned block), in which case a DOM-accurate List View will be crucial to make sense of the whole.

**Reply by jameskoster** [reply]
As a designer I am always considering the needs of users :) With most of these controls, we need to make performing common actions quick and easy, but not at the expense of more granular control. "I want to make my header sticky" is a different proposition to "I want this one block to be fixed in position and not scroll". We need solutions that work for both. If we get the foundational elements right first, we can build on those to provide solutions for more specific use cases.

I'm hearing List View will be tricky, so we can look into it separately.

**Reply by tellthemachines** [reply]
> As a designer I am always considering the needs of users :)

😄 yeah, that phrasing was probably a bit too vague. What I meant was to consider it functionality-first (what controls to expose and what they should do) as opposed to implementation-first (e.g. which of the CSS position values we should expose), because implementation can vary according to our needs.

In this case, I'm wondering the following:

* when making a block sticky, should it default to be sticky throughout the whole scrollable page, unless we explicitly set an area for it to be sticky within? 
* What would a control for setting the "sticky area" look like?
* similarly with absolute positioning, are we positioning blocks page-wide, or will we allow an explicit container area to be set?

I think fixed is likely to be the most straightforward in terms of controls, because it will always be relative to the viewport. Perhaps we can start with that one?

**Reply by andrewserong** [reply]
Good questions!

> when making a block sticky, should it default to be sticky throughout the whole scrollable page, unless we explicitly set an area for it to be sticky within?

I know it's likely confusing for users, but I quite like the browser default of sticky applying to the height of the direct parent of the currently selected block, as it allows for some quite creative designs (e.g. intentionally making a group block sticky for only part of a page, within a column etc), but I agree it's a really challenging one to communicate clearly in terms of usability 🤔 

> What would a control for setting the "sticky area" look like?

That's a really interesting idea. If we tried a control for selecting a sticky area, I'm also wondering how would it work? For an initial implementation of sticky/fixed, I'd love it if we can avoid depending on shipping JS to the site frontend, since we don't have a precedent for that within the block supports, so it could be a bit of a rabbithole.

> I think fixed is likely to be the most straightforward in terms of controls, because it will always be relative to the viewport. Perhaps we can start with that one?

I like the idea of starting with whatever's going to be the simplest UX, but unfortunately so far, I've found `fixed` a bit harder to work with. With `fixed` we run into some UX / UI challenges with ensuring the correct positioning and width, and that there is enough padding for the space that the element would have otherwise taken up (browser default behaviour for `sticky` deals with the width and padding automatically). Here's [a quick codepen](https://codepen.io/andrewserong/pen/abKKgPW) for exploration:

| Browser default sticky (note the the following paragraph is visible, the sticky `div` has full width) | Browser default fixed (note the paragraph is hidden behind the fixed element, the `div` is the width of its content) |
| --- | --- |
| <img width="944" alt="image" src="https://user-images.githubusercontent.com/14988353/204223088-31f587d6-dc4b-4842-9681-ccd9a98d8ac8.png"> | <img width="943" alt="image" src="https://user-images.githubusercontent.com/14988353/204223301-a49801b4-4bae-4077-8192-fa60e84eff68.png"> |

Because of that, out-of-the-box `sticky` appears to get us closer to user-friendly behaviour in terms of the block behaving as it's expected to, but we run into the challenge of it being a hard concept to work with (parent / child relationship and spacing), whereas with `fixed` the controls might behave in more of a directly obvious way, but we then have a challenge of figuring out how to handle the spacing in a programatic way that works with the site layout. 🤔

Hrm... now that we're discussing the details of the trade-offs, I think both approaches are going to be pretty challenging! I think I'd lean a little more toward starting with `sticky`, since it doesn't require width controls or calculating and outputting site padding to account for it. It'd be great to be able to support both, though.

**Reply by jameskoster** [reply]
I think the most important thing to remember is that we are essentially creating a UI for CSS, so at the most foundational layer the controls should reflect the natural behaviour therein. So yes, a 'sticky' block should 'unstick' when its container scrolls out of view. It's also worth noting that this is a fairly advanced control, and non-technical users less likely to interact with it – more likely they will encounter its effects as a part of theme / pattern selection. Once we get the foundational layer right we can explore creating UIs for dedicated use cases.

**Comment by jameskoster** [comment]
One other option we might consider for the `fixed` position would be to permanently display the position picker, which emphasises the simpler way to set position at the expense of some visual real estate:

<img width="1257" alt="Screenshot 2022-11-25 at 13 29 04" src="https://user-images.githubusercontent.com/846565/203995808-95a3b21e-3a70-4587-ad95-b54d7b139a3c.png">


As before we could do the same thing for the `absolute` positioning option.

**Comment by jasmussen** [comment]
I'm coming in late to this one, sorry was under the weather last week. Lots of good thoughts here, lots of excellent mockups. A few quick thoughts:

* I don't think the goal should be to reimplement CSS. But ultimately that's still what we're bound by. We can find a good UI for it by starting with the most salient pieces, implementing them well, balancing prominence, and providing good patterns and templates for these.
* In that vein, I'd omit relative and static for now as explicit choices. Static can be implied as a default/unset, and relative would be interesting to see if we could do without, but add if we need to.
* As far as the 9-grid, I like that interface a lot. It provides smart handy defaults for the majority of positions (though I'd omit the one in the center, making it an 8 grid). 
* The settings toggle for custom inputs working like it does for font size is smart. But I'd make it a toggle instead of the popover, so the grid is _replaced_ by the inputs. That way we don't need to highlight your custom position once you close the popover.
* I don't think "Advanced" is the right place for this. I still think Layout is, and so long as it's a non-default ToolsPanel control, it manages its prominence there.

Here's a quick things at taking Jay's ingredients and tweaking them a bit:

<img width="2390" alt="Sticky toggle" src="https://user-images.githubusercontent.com/1204802/204295540-71a505ab-5bb0-462e-a1c5-cf1e412922de.png">

These actually integrate another idea of Jay's, which is to integrate a "quick shortcut" dropdown inside the input. It's like the unit picker, a sort of shorthand just like the unit control. What do you think?

**Reply by jameskoster** [reply]
> I don't think "Advanced" is the right place for this. I still think Layout is, and so long as it's a non-default ToolsPanel control, it manages its prominence there.

My suggestion to move the control to Advanced is less about prominence and more about context. _All_ other controls in the Layout panel relate specifically to the containers _children_, whereas position is applied to the container itself. 

There's also the issue of non-container blocks – you might want to add 'position' to something like a paragraph, but that block does not have a Layout panel. Do we really want to add one only for this affordance? It seems a shame to me. Is there anywhere else to put this control?

> tweaking them a bit

Thanks for sharing. I like how the control appears simpler, but have a couple of concerns:

1. It looks like a text input – should it be more obvious that it's a dropdown?
2. I appreciate the use of the quick shortcut pattern, but this implementation is actually inconsistent with the selector we've discussed for width. There the icon is the element that switches the type, and the input contains the value. If we wanted to be consistent it would be more like this:

<img width="673" alt="Screenshot 2022-11-28 at 15 00 45" src="https://user-images.githubusercontent.com/846565/204310044-a79d84cf-3522-422a-8141-1d451ce5350d.png">

This suffers from the text-input-issue I mentioned above though, and it gets a bit complicated when you think about the toggle to display the input values. This isn't great:

<img width="829" alt="Screenshot 2022-11-28 at 15 03 07" src="https://user-images.githubusercontent.com/846565/204310647-ab71c668-c2fc-4e74-b29e-304923008098.png">

**Reply by jasmussen** [reply]
Indeed, this is a tricky one. I think I'll need to clear my head a bit 😅 — [here's the Figma](https://www.figma.com/file/vmHxwa8WhiyiP9poe2om9J/Inspector-Panels?node-id=13565%3A53143&t=TLVn9lHvakMTcQB6-1) for the time being.

**Reply by andrewserong** [reply]
I don't mind the move to Advanced personally, and think it could be a good initial step, but just adding a note about this comment:

> I don't think "Advanced" is the right place for this. I still think Layout is, and so long as it's a non-default ToolsPanel control, it manages its prominence there.

We have a WIP PR to explore switching the LayoutPanel over to the ToolsPanel over in #45833, however that exploration resulted in more discussion about rethinking how the LayoutPanel works, its included controls, and default display etc, which seemed to increase the scope for making changes quite a bit (see discussion at: https://github.com/WordPress/gutenberg/pull/45833#issuecomment-1326271441). I'm still keen for us to pursue improvements to the panel, but it sounded like in the shorter-term that PR wasn't going to be viable due to the number of controls / design issues to be thought through?

I'd be in favour of seeing if we can initially have the position control in Advanced. Though we should probably keep in mind that the Advanced panel and its slot also do not use the ToolsPanel, and converting to ToolsPanel is likely not going to be easy there as plugins already use the Advanced slot. So if we place Position in Advanced, we might need to accept (at least in the short-term) that it's still the accordion-like single show/hide panel for that whole group of controls.

**Reply by andrewserong** [reply]
To see how it feels in practice to place the control under Advanced, I've opened up an alternate sticky position PR that we can play with here: https://github.com/WordPress/gutenberg/pull/46142

**Reply by jasmussen** [reply]
So I think we can still start with Sticky, and expand to Fixed at a later time. The key bit is we need to have an interface that more naturally scales to that, than the toggle could. In that light, here's another attempt. The near term could be this dropdown:

<img width="1466" alt="Sticky toggle, near" src="https://user-images.githubusercontent.com/1204802/204483282-c5e0e6b5-7b1d-4f34-84de-7cbb7989892b.png">

Not ideal, but at least the dropdown would scale to Fixed and anything else we wanted. And so the longer term could be this:

<img width="2488" alt="Sticky toggle, long term" src="https://user-images.githubusercontent.com/1204802/204483381-614d9254-5ad4-48a4-9337-6281cfcc772a.png">

That is, the split input is inspired by the border control, and opens the position picker leftwards.

What do you think?

**Reply by andrewserong** [reply]
> The key bit is we need to have an interface that more naturally scales to that, than the toggle could.

Nice work, with each of these iterations it feels like we're honing in on a good direction to me, and I like the idea of starting with the dropdown and just static / sticky to begin with 👍. I also quite like how compact the control looks with the split control, and the custom button for opening up the granular offsets for top/right/bottom/left, too (looks similar to the border radius unlinked controls). So a plus one from me for starting with a dropdown and scaling from there.

One question: these latest designs still depict a Layout panel that has been converted to the ToolsPanel, and the control appearing there. Is that a requirement to begin with? Let me know if you think we should revive the ToolsPanel conversion at this stage (the work over in https://github.com/WordPress/gutenberg/pull/45833), or if we can begin with the Advanced panel.

**Reply by jasmussen** [reply]
> One question: these latest designs still depict a Layout panel that has been converted to the ToolsPanel, and the control appearing there. 

@jameskoster and I discussed yesterday approaches here, and the question of the Layout panel came up, but perhaps more importantly, the question of whether this feature should be available for _all_ blocks. I.e. should it be possible to make a paragraph sticky?

I think as we figure out where to ultimately put it (looks like the split inspector experiment might make it feasible to add this to all blocks), it would be fine to put it in advanced so far, as Jay has suggested in the past. Though I'd keep advanced a closed-by-default panel for that. 

I think the bigger question to figure out before moving with the near-term solution, really is whether this is a thing we apply to all blocks or just a few. The safest approach might be to _assume_ we want it on every block, but start with just Group, Row, Stack blocks and then expand the range as we learn more. It would be harder if we started with the feature available for every block, and then pulled back. Would that be feasible, and what do you think?

**Reply by jameskoster** [reply]
If we don't want to use the toggle due to concerns about changing it down the road, then don't we need to get the long-term vision for this control right before merging anything? In that respect I don't think we're quite there, lots of questions still need to be answered:

* Can you position any block or only containers?
* Do we need other position options in the future (`relative`, `absolute`), and if so how does the control handle them?
* Is there a way to omit the default (static) position in the design?
* In the granular controls, the Top, Right, Bottom, and Left inputs will need to entertain the default `auto` value, how does that look/work?
* With the grid picker, a selection like top, center requires additional behind-the-scenes css magic to make it work (`left: 50%; transform: translateX(-50%);`). How is that exposed when you open the granular control? What about `z-index`?
* How does Position interact with global styles? Can you globally declare a block type is always Sticky?
* How can we abstract all this complexity for really simple use cases ("make my site header sticky").

Some of these questions are bigger than others, but I don't think they're insurmountable. It just needs a little more time in the design oven, imho. Once we get everything right we can row back a bit and figure out the short term next steps.

**Reply by jasmussen** [reply]
If we could find some compelling answers to those excellent questions, would we be able to have something demoable in a PR in the next couple of weeks? That is, demoable even with rough edges and limitations. I think that's a technical question, and the design question then becomes: what can we shave off to make it more likely to be possible to show something 🤔

**Reply by andrewserong** [reply]
+1 these are all excellent questions, and I agree about taking the time to make sure the direction is right, rather than try to rush through something that might lock us in prematurely 👍. A couple of potential answers to some of the questions that come immediately to mind:

> Can you position any block or only containers?

For sticky / fixed, I think it'll work most reliably with containers, so I like Joen's suggestion of starting with Group, Row, Stack and expanding from there. I'm currently building it as an opt in block support, so that we can deliberately opt in as needed, and that could potentially expand to as many blocks as we like (much like opting in to margin/padding).

> Is there a way to omit the default (static) position in the design?

Whether or not we call it default or static, it should be easy for a user to see how to switch off the position setting. For the moment, having a default/static option sounds like a good way to do it to me.

> How does Position interact with global styles? Can you globally declare a block type is always Sticky?

Personally, I think it should only be set at the individual block instance level, and shouldn't have any global styles support. The use case for setting a position globally for a particular block type is probably not very common, and it would be very difficult to get it work in a way that feels predictable.

**Reply by andrewserong** [reply]
> If we could find some compelling answers to those excellent questions, would we be able to have something demoable in a PR in the next couple of weeks? That is, demoable even with rough edges and limitations.

We sure can have _something_ demoable — there's a working PR over in https://github.com/WordPress/gutenberg/pull/46142 that has the previous toggle-sticky control set under the Advanced panel (this is the PR I'm actively working on right now). I'm happy to keep chipping away at that PR over the next couple of weeks and try to add in as much as we decide on based on the discussions here (or at least make a start on it). My goal is to keep that PR in a working state, so that it can be played with at any point.

The main caveat of demoing the WIP is that I'm sure we'll run into a bunch of currently unknown hurdles, and potential roadblocks (e.g. the toolbar positioning issues begin worked on in: https://github.com/WordPress/gutenberg/pull/46085). So, provided that anyone looking knows that even if it looks close to good/done, it most likely isn't 😅, I think having usable demo PRs is a great idea!

_Update: I thought I'd do some more hacking, and I've had a go in the [WIP PR](https://github.com/WordPress/gutenberg/pull/46142#issuecomment-1333270628) of trying out a custom select control for the dropdown with static/default and sticky options. There's a screengrab on the WIP PR here: https://github.com/WordPress/gutenberg/pull/46142#issuecomment-1333270628._

**Reply by jasmussen** [reply]
Honestly, https://github.com/WordPress/gutenberg/pull/46142#issuecomment-1333270628 looks like a fair enough start to me. 

* We seem to be settling on a dropdown for the longer run, even if it might come with additional contextual properties.
* We're not sure where to place it, but we know it's advanced, so putting it in the collapsible advanced section is fine.
* We're not sure whether to apply it to only a few blocks, or all, so starting with containers so long as we _can_ expand, feels pretty safe.

I personally think this is a good start. It's not a strong opinion. I think the only concern with this approach would be that the control will most likely _move_ from "Advanced" into something else. But outside of coming up with an excellent existing place for it, figuring out new panels seems like it should be a more holistic exercise. What do you think, @WordPress/gutenberg-design?

**Reply by jameskoster** [reply]
I'd like to share another concept that I think can be implemented today, and scale elegantly into the future as we add support for more position options:

### Short term

<img width="957" alt="Screenshot 2022-12-01 at 10 28 20" src="https://user-images.githubusercontent.com/846565/205029473-477e60c5-4415-467c-990c-ec8ba0039322.png">


Position is a dedicated panel in the Settings tab. It's familiar to anyone who knows CSS, and thanks to the split inspector tabs the prominence seems appropriate. As a setting, it will not appear in global styles. This format gets around the need to have a default (`static`) positional option. In this case the absence of position = static.

Fixed position only offers the grid selector to begin with. This enough to solve most common use cases (banners, fixed footer, etc).

Sticky has no options, and essentially just adds `position: sticky; top: 0;`. Again this is enough to solve common use cases (Sticky header, Table of Contents, Alphabetised list, etc).

### Long term
<img width="920" alt="Screenshot 2022-12-01 at 10 21 55" src="https://user-images.githubusercontent.com/846565/205028152-1510dd8d-924a-470c-9c1d-19f45fc8a95f.png">

Down the road it's easy to expand the panel to accommodate more granular controls, and other options like relative and absolute positioning.

**Reply by jasmussen** [reply]
I think there's good sense in a position panel, especially if the split inspector graduates from experimental. I also like the simplicity of the positional selection interface. The geograpically placed position values looks good too, but could probably be closer to the border control:

<img width="315" alt="image" src="https://user-images.githubusercontent.com/1204802/205031229-5990985f-fed0-4b41-8795-8cba24d3d956.png">

I'm also not yet sure if position relative is a property we _want_ to surface — I'd love to see how far we could get with assuming all containers are simply relative, mainly for the reason that the use case of absolute position blocks breaking out of their nearest container seems rare. To be sure, we can add it if need be, but it seems sensible to try and start without it.

But I have a small concern that it departs in behavior from existing toolspanels. It's effectively a new panel style, which would not be trivial for the near term. I realize that a position toolspanel with only one option is kind of useless, but would a collapsible panel be better there? Though I've argued against advanced as the placement, it does seem preferrable to have it there for the near term.

**Reply by jameskoster** [reply]
> I realize that a position toolspanel with only one option is kind of useless, but would a collapsible panel be better there?

It's not much consolation but it wouldn't be a single option. If we use a collapsible panel then we'll be back in the position of having to surface a 'Default' value. It might work?

<img width="1050" alt="Screenshot 2022-12-01 at 11 39 20" src="https://user-images.githubusercontent.com/846565/205043747-a58d2c7d-f2b9-4835-906a-c556efc154c8.png">

**Reply by jasmussen** [reply]
That might work, yes, but it would still be dependant on the split inspector graduating, at least if we were to roll it out beyond the Group. But if we did limit it to that, then maybe a position panel might be okay 🤔 

About the inline positional picker. How would you feel about finding a way to make it into a popover? It takes up a fair bit of space, and I think there's some sense in reusing the pattern from the Cover matrix button, which is a popover as well.

**Reply by jameskoster** [reply]
> That might work, yes, but it would still be dependant on the split inspector graduating

I guess this design could go in the Advanced panel as well. But alongside everything else in there the usability degrades quite a bit imo. Position is complex and warrants its own space.

> About the inline positional picker. How would you feel about finding a way to make it into a popover? 

In this context I feel that pattern would put too much emphasis on the dropdown. After selecting that you want a fixed position, the dropdown effectively becomes secondary. If we used the accordion panel you can always just collapse it to reduce unwanted distraction. 

There might be other options though, like a two column layout: 

<img width="284" alt="Screenshot 2022-12-01 at 14 44 30" src="https://user-images.githubusercontent.com/846565/205082158-ee2524dd-125f-49a2-a8d7-8f9c99ba32ed.png">

**Reply by andrewserong** [reply]
> But I have a small concern that it departs in behavior from existing toolspanels. It's effectively a new panel style, which would not be trivial for the near term.

I have the same concern (similar to when the concept was suggested for the layout panel in https://github.com/WordPress/gutenberg/pull/45833#issuecomment-1324553833) in that it introduces a complexity to the concept of the panel, where the plus icon becomes an action that affects block state, along with the minus icon that removes block state. While the UI looks fairly simple, this isn't something supported by the ToolsPanel and introduces a different kind of interaction with panels than what we already have in the editor UI, which opens up some accessibility challenges. E.g. what does it mean to add a position, how does a user know what they're about to do before they click on the plus icon when in other cases the plus icon is a drop-down menu that exposes options.

> If we use a collapsible panel then we'll be back in the position of having to surface a 'Default' value. It might work?

I quite like the collapsible panel idea for position, though, and it side-steps the problem of creating a new panel type. I'd be happy to try doing that.

Personally, I'm still in favour of having a 'Default' value in the UI so that a user can quickly see what the current state is, rather than inferring it based on the absence of a control being visible. In that respect it's sort of similar to how the font-size dropdown displays 'Default' when no font size is selected (on themes with lots of font-sizes). What's the reasoning behind wanting to hide it?

**Reply by andrewserong** [reply]
Update: for fun, over in the WIP PR, I've had a go at moving the control over to a separate collapsible Position panel. Seems to work pretty well so far: https://github.com/WordPress/gutenberg/pull/46142#issuecomment-1334785758

**Comment by simison** [comment]
While your use-case is clearly focused more on the header, would this also allow creating UIs for floating modals and things like cookie banners?

Those are trickier than header because there should be a way to hide them for most of the time so that they aren't obscuring the view all the time.

**Reply by jameskoster** [reply]
> While your use-case is clearly focused more on the header

Actually we've tried to design all the concepts here to be block-agnostic. 

Modals are a special case due to the requirement of trigger / dismissal components, there's a lot more to it than just positioning. We can perhaps look at the Navigation block's Overlay option for inspiration. When we get there, the position controls we're working on here should contribute effectively as a part of any modal creation flow, but can probably be de-emphasised a bit since you'd generally want modals to be fixed / center aligned by default.

**Comment by jasmussen** [comment]
One aspect of positioning presets is to find good values to apply for each preset. It might be useful to explore exactly what such presets might look like, so I took a stab in this codepen: https://codepen.io/joen/pen/wvXLORR?editors=1100

<img width="626" alt="Screenshot 2022-12-14 at 15 18 09" src="https://user-images.githubusercontent.com/1204802/207619526-387b1c08-c90e-40e2-a2d3-2b2030b0c7c2.png">

It has 9 preset positions, and each of them are applied by combining two classes, such as "is-top is-left". This involves not just top/left/right/bottom values, but for middle and center positions, it needs a little translate as well. Is that feasible? Are there good alternatives, maybe `calc`-based values?

**Reply by andrewserong** [reply]
> Is that feasible?

Conceptually the way the WIP work has been explored so far, we should be able to output additional rules. We'll likely need to add the transform properties to the allowed list for `kses`, which we need to do for `top/right/bottom/left` anyway, so I think it's probably feasible. I also quite like the utility classnames for these preset-like values, too 👍 I suppose we'll still want to support direct values (at least technically) for designs that want a deliberate offset (e.g. `2em` from top, or `2em` from bottom, etc), but this looks promising to me.

**Reply by jasmussen** [reply]
> I suppose we'll still want to support direct values (at least technically) for designs that want a deliberate offset (e.g. 2em from top, or 2em from bottom, etc)

Yes, we'll want to support manual values. Manually _centering_, is going to be a challenge, the translate hack isn't easy to grok. But it's a challenge we can meet, I'm sure.

**Reply by jameskoster** [reply]
> Manually centering, is going to be a challenge

If we consider the user flows I think it might be ok. 

In most cases I suspect folks would select MC using the preset which provides the context they need to interpret the controls that are revealed when switching to manual config. An important part of this being that the controls necessary to achieve the preset are revealed (and populated) when you make that switch. 

Going directly to manual config without choosing a preset is more advanced, and we can perhaps assume that these users understand the different positioning methods. Still, we can potentially hide the transform control by default and make it possible to add via the tools panel.

**Reply by jasmussen** [reply]
Perhaps we can get very far with these 9 presets, and then top/right/bottom/left manual controls 🤔

**Comment by Faton1995** [comment]
Hello, is it not possible to integrate this in the following way: When a block is fixed, it remains fixed at the same position as it is displayed in the full site editor?
